++ in local mode broken ?

```
  import com.twitter.scalding._

  class Broken(args : Args) extends Job(args) {

   val data = for {
     row <- (1 to 10).toList
     col <- List(1,2)
   } yield (row,col)

   // This works
   //IterableSource(data, ('a,'b)).read).write(Tsv("sum"))

   // This does not
   (IterableSource(data, ('a,'b)).read ++ IterableSource(data, ('a,'b)).read).write(Tsv("sum"))
 }
```

---

[tw-mbp13-kraman kraman (invsg)]$ scald.rb --local Broken.scala
[INFO] Found Job Class: Broken
compiling Broken.scala
scalac -classpath /var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/maven/hadoop-core-0.20.2.jar:/var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/maven/log4j-1.2.15.jar:/var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/maven/commons-httpclient-3.1.jar:/var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/maven/commons-cli-1.2.jar:/var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/maven/zookeeper-3.3.4.jar:/Users/kraman/workspace/science/dist/ads-batch-deploy.jar -d /var/folders/b_/17q0nsss269_2kf855mtg4_c0000gn/T/script-build Broken.scala
Exception in thread "main" java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:39)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:27)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:513)
    at com.twitter.scalding.Job$.apply(Job.scala:33)
    at com.twitter.scalding.Tool.getJob(Tool.scala:51)
    at com.twitter.scalding.Tool.run(Tool.scala:72)
    at org.apache.hadoop.util.ToolRunner.run(ToolRunner.java:65)
    at com.twitter.scalding.Tool$.main(Tool.scala:128)
    at com.twitter.scalding.Tool.main(Tool.scala)
Caused by: java.lang.NullPointerException
    at java.util.Arrays$ArrayList.<init>(Arrays.java:3357)
    at java.util.Arrays.asList(Arrays.java:3343)
    at cascading.pipe.Splice.<init>(Splice.java:274)
    at cascading.pipe.Splice.<init>(Splice.java:226)
    at cascading.pipe.Splice.<init>(Splice.java:186)
    at cascading.pipe.Splice.<init>(Splice.java:175)
    at cascading.pipe.Splice.<init>(Splice.java:422)
    at cascading.pipe.Merge.<init>(Merge.java:49)
    at com.twitter.scalding.RichPipe.$plus$plus(RichPipe.scala:165)
    at Broken.<init>(Broken.scala:14)
##     ... 10 more

--- a/scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala
+++ b/scalding-core/src/main/scala/com/twitter/scalding/RichPipe.scala
@@ -173,7 +173,16 @@ class RichPipe(val pipe : Pipe) extends java.io.Serializable with JoinAlgorithms
   /**
    * Merge or Concatenate several pipes together with this one:
    */
-  def ++(that : Pipe) = new Merge(assignName(this.pipe), assignName(that))
+  def ++(that : Pipe): Pipe = {
+    if(this.pipe == that) {
+      // Cascading fails on self merge:
+      // solution by Jack Guo
+      new Merge(assignName(this.pipe), assignName(new Each(that, new Identity)))
+    }
+    else {
+      new Merge(assignName(this.pipe), assignName(that))
+    }
+  }
 
   /**
    * Group all tuples down to one reducer.
@@ -315,7 +324,7 @@ class RichPipe(val pipe : Pipe) extends java.io.Serializable with JoinAlgorithms
     conv.assertArityMatches(f)
     new Each(pipe, f, new FilterFunction(fn, conv))
   }
-  
+
   /**
    * Text files can have corrupted data. If you use this function and a
    * cascading trap you can filter out corrupted data from your pipe.

--- a/scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala
+++ b/scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala
@@ -373,8 +373,17 @@ final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) ext
   override def fork: TypedPipe[T] =
     MergedTypedPipe(left.fork, right.fork)
 
-  def toPipe[U >: T](fieldNames: Fields)(implicit setter: TupleSetter[U]): Pipe =
-    new cascading.pipe.Merge(left.toPipe[U](fieldNames), right.toPipe[U](fieldNames))
+  def toPipe[U >: T](fieldNames: Fields)(implicit setter: TupleSetter[U]): Pipe = {
+    if(left == right) {
+      //use map:
+      left.flatMap {t => List(t, t)}.toPipe[U](fieldNames)
+    }
+    else {
+      import RichPipe.assignName
+      new cascading.pipe.Merge(assignName(left.toPipe[U](fieldNames)),
+        assignName(right.toPipe[U](fieldNames)))
+    }
+  }
 }
 
 class TuplePipeJoinEnrichment[K, V](pipe: TypedPipe[(K, V)])(implicit ord: Ordering[K]) {

--- a/scalding-core/src/test/scala/com/twitter/scalding/CoreTest.scala
+++ b/scalding-core/src/test/scala/com/twitter/scalding/CoreTest.scala
@@ -555,6 +555,9 @@ class MergeTestJob(args : Args) extends Job(args) {
   val small = in.filter('x) { (x:Double) => (x <= 0.5) }
   (big ++ small).groupBy('x) { _.max('y) }
   .write(Tsv(args("out")))
+  // Self merge should work
+  (big ++ big).groupBy('x) { _.max('y) }
+  .write(Tsv("out2"))
 }
 
 class MergeTest extends Specification {
@@ -579,6 +582,11 @@ class MergeTest extends Specification {
       sink[(Double,Double)](Tsv("fakeOutput")) { outBuf =>
         "correctly merge two pipes" in {
           golden must be_==(outBuf.toMap)
+        }
+      }.
+      sink[(Double,Double)](Tsv("out2")) { outBuf =>
+        "correctly self merge" in {
+          outBuf.toMap must be_==(big.groupBy(_._1).mapValues{iter => iter.map(_._2).max})
         }
       }.
       run.
@@ -1580,12 +1588,12 @@ class SampleWithReplacementJob(args : Args) extends Job(args) {
 
 class SampleWithReplacementTest extends Specification {
   import com.twitter.scalding.mathematics.Poisson
-  
+
   val p = new Poisson(1.0, 0)
   val simulated = (1 to 100).map{
     i => i -> p.nextInt
   }.filterNot(_._2 == 0).toSet
-  
+
   noDetailedDiffs()
   "A SampleWithReplacementJob" should {
     JobTest("com.twitter.scalding.SampleWithReplacementJob")
@@ -1611,11 +1619,11 @@ class VerifyTypesJob(args: Args) extends Job(args) {
 }
 
 class VerifyTypesJobTest extends Specification {
-  "Verify types operation" should { 
+  "Verify types operation" should {
     "put bad records in a trap" in {
            val input = List((3, "aaa"),(23,154),(15,"123"),(53,143),(7,85),(19,195),
              (42,187),(35,165),(68,121),(13,"34"),(17,173),(2,13),(2,"break"))
-           
+
            JobTest(new com.twitter.scalding.VerifyTypesJob(_))
              .source(Tsv("input", new Fields("age", "weight")), input)
              .sink[(Int, Int)](Tsv("output")) { outBuf =>
@@ -1626,7 +1634,7 @@ class VerifyTypesJobTest extends Specification {
              }
              .run
              .finish
-             
+
          }
   	}
 }

--- a/scalding-core/src/test/scala/com/twitter/scalding/TypedPipeTest.scala
+++ b/scalding-core/src/test/scala/com/twitter/scalding/TypedPipeTest.scala
@@ -504,3 +504,33 @@ class TypedFlattenTest extends Specification {
       .finish
   }
 }
+
+class TypedMergeJob(args: Args) extends Job(args) {
+  val tp = TypedPipe.from(TypedTsv[String]("input"))
+  (tp ++ tp)
+    .write(TypedTsv[String]("output"))
+  (tp ++ (tp.map(_.reverse)))
+    .write(TypedTsv[String]("output2"))
+}
+
+class TypedMergeTest extends Specification {
+  import Dsl._
+  noDetailedDiffs()
+  "A TypedMergeJob" should {
+    JobTest(new TypedMergeJob(_))
+      .source(TypedTsv[String]("input"), List(Tuple1("you all"), Tuple1("every body")))
+      .sink[String](TypedTsv[String]("output")) { outBuf =>
+        "correctly flatten" in {
+          outBuf.toSet must be_==(Set("you all", "every body"))
+        }
+      }
+      .sink[String](TypedTsv[String]("output2")) { outBuf =>
+        "correctly flatten" in {
+          val correct = Set("you all", "every body")
+          outBuf.toSet must be_==(correct ++ correct.map(_.reverse))
+        }
+      }
+      .runHadoop
+      .finish
+  }
+}

